bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/TypeChecker.java

```diff
@@ -1156,17 +1196,24 @@ private static boolean checkRecordBelongsToAnydataType(MapValue sourceVal, BReco
 
     private static boolean checkIsRecordType(Object sourceVal, Type sourceType, BRecordType targetType,
                                              List<TypePair> unresolvedTypes) {
-        if (sourceType.getTag() != TypeTags.RECORD_TYPE_TAG) {
-            return false;
+        switch (sourceType.getTag()) {
+            case TypeTags.RECORD_TYPE_TAG:
+                return checkIsRecordType((MapValue) sourceVal, (BRecordType) sourceType, targetType, unresolvedTypes);
+            case TypeTags.MAP_TAG:
+                return checkIsRecordType((BMapType) sourceType, targetType, unresolvedTypes);
```

Comment:
Is it enough to simply check if the two types are compatible? What about the behaviour if the record type contains a `never` field? :thinking:

Reply:
Since we check the source map's constraint against the field type in https://github.com/ballerina-platform/ballerina-lang/pull/26533/files#diff-19c04b50f4303ca57559920287e2ca9ef20fd7ddaad5b7c75d425c340ec5421dR1151, that shouldn't be an issue, right?

Reply:
Had a chat with Maryam and got this clarified.